iothub/device/src/Pipeline/ClientTransportStateMachine.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Devices.Client.Transport
+{
+    internal enum ClientTransportState
+    {
+        Closed,
+        Opening,
+        Open,
+        Closing,
+    }
+
+    internal enum ClientStateAction
+    {
+        OpenStart,
+        OpenSuccess,
+        OpenFailure,
+        CloseStart,
+        CloseComplete,
+        ConnectionLost,
+    }
+
+    internal class ClientTransportStateMachine
```

Comment:
Obviously you aren't there yet, but you'll be adding copious code comments and some unit tests, yes?